mantidproject/mantid
# Replace Engineering Diffraction Calibration and Focussing Python workflows with C++ workflows

**Description of work.**

Calibration and Focussing (in both the gui and EnginX scripts) made use of slow and unreliable python algorithms. This PR replaces those workflows with CPP algorithm based workflows (PDCalibration).
Some other miscellaneous changes: also output focussed runs in dSpacing as well as TOF



**To test:**


1. Using the Engineering Diffraction gui, create a calibration (e.g vanadium 236516 ceria 193749)
2. Also create a calibration for a single bank, or for a crop of spectrum numbers (try 34-678 or something)
3. Focus a sample run using that calibration (e.g. 299080)
4. Check outputs on bank calibration to see if the peaks have a good agreement (fairly well aligned by dSpacing)
5. Give the EnginX script example a run with the same run number inputs and see if the outputs align between this and the gui

```from Engineering.EnginX import main

path = '/change/this/path/to/whole_inst_calib.nxs'
main(vanadium_run='236516', user='0110', focus_run='299080', ceria_run='193749', full_inst_calib_path=path)

# Kwargs:
#         ceria_run (string): the run number of the ceria to use
#         force_vanadium (bool): forces creation of vanadium even if one for the run already exists
#         force_cal (bool) : forces creation of calibration files
#         full_inst_calib_path (string) : path to the full instrument calibration file used in calibration and focussing
#         crop_type (string): what method of cropping to use
#         crop_name (string): what to call the cropped bank workspace
#         crop_on (string): the bank of spectrum to crop on if cropping
#         pre_process_run (bool): set whether or not to pre-process run before focusing
#         params (string): rebin parameters for pre-process
#         time_period (string): time period for pre-process
#         grouping_file (string): the path of the grouping file for texture focusing
#         directory (string): the path of the directory to save to
#         user_struct (bool): whether or not to use the enginx file structure (defaults to true)
```

Full instrument calibration file (needed for both the gui and scripts):
[full_inst_calib.zip](https://github.com/mantidproject/mantid/files/6473094/full_inst_calib.zip)


Fixes #30048 and Fixes #30288 




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
